uvtools/dspec.py

```diff
@@ -146,6 +147,250 @@ def delay_filter(data, wgts, bl_len, sdf, standoff=0., horizon=1., min_dly=0.0,
+    
+    assert(len(cn.shape) == 1), "cn must be a 1D array"
+    nmax = (cn.size - 1) // 2 # Max. harmonic
+    
+    # Build matrix operator for complex Fourier basis
+    n = np.arange(-nmax, nmax+1)
+    F = np.array([np.exp(-1.j * _n * nu / L) for _n in n])
+    
+    # Return model
+    return np.dot(cn, F)
+
+
+def delay_filter_leastsq_1d(data, flags, sigma, nmax, add_noise=False, 
+                            cn_guess=None, use_linear=True, operator=None):
+    """
+    Fit a smooth model to 1D complex-valued data with flags, using a linear 
+    least-squares solver. The model is a Fourier series up to a specified 
+    order. As well as calculating a best-fit model, this will also return a 
+    copy of the data with flagged regions filled in ('in-painted') with the 
+    smooth solution. 
+    
+    Optionally, you can also add an uncorrelated noise realization on top of 
+    the smooth model in the flagged region.
+    
+    Parameters
+    ----------
+    data : array_like, complex
+        Complex visibility array as a function of frequency, with shape 
+        (Nfreqs,).
+    
+    flags : array_like, bool
+        Boolean flags with the same shape as data.
+    
+    sigma : float or array_like
+        Noise standard deviation, in the same units as the data. If float, 
+        assumed to be homogeneous in frequency. If array_like, must have 
+        the same shape as the data.
+        
+        Note that the choice of sigma will have some bearing on how sensitive 
+        the fits are to small-scale variations.
+    
+    nmax: int
+        Max. order of Fourier modes to fit. A model with complex Fourier modes 
+        between [-n, n] will be fitted to the data, where the Fourier basis 
+        functions are ~ exp(-i 2 pi n nu / (Delta nu).
+    
+    add_noise : bool, optional
+        Whether to add an unconstrained noise realization to the in-painted areas. 
+        This uses sigma to set the noise standard deviation. Default: False.
+    
+    cn_guess : array_like, optional
+        Initial guess for the series coefficients. If None, zeros will be used. 
+        A sensible choice of cn_guess can speed up the solver significantly. 
+        Default: None.
+    
+    use_linear : bool, optional
+        Whether to use a fast linear least-squares solver to fit the Fourier 
+        coefficients, or a slower generalized least-squares solver. 
+        Default: True.
+    
+    operator : array_like, optional
+        Fourier basis operator matrix. Must have shape (Nfreq, Nmodes), where 
+        Nmodes = 2*nmax + 1.
+    
+    Returns
+    -------
+    model : array_like
+        Best-fit model, composed of a sum of Fourier modes.
+        
+    model_coeffs : array_like
+        Coefficients of Fourier modes, ordered from modes [-n, +n].
+    
+    data_out : array_like
+        In-painted data.
+    """
+    # Construct Fourier basis operator if not specified
+    if operator is None:
+        # Frequency array
```

Comment:
This block of code is used in the function above too. It might make sense to make the basis generation its own function, to make sure it will be the same for both uses. (I'm assuming that the average user will call this function without specifying their own basis, and then want the option to use `fourier_model` to generate an actual spectrum from just the saved coefficients. If you had a different use case in mind, then this comment might be less relevant.)

Reply:
Done.